hips/hip-9999.md

```diff
@@ -0,0 +1,92 @@
+created: "2022-02-01"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+Composing Helm charts using dependencies is currently limited because it is not possible to create a chart with a defined set of values and distribute them to subcharts. Users of a parent chart instead have to provide values to the subcharts directly. Alternatively, users can provide `global` values, but in that case the subcharts need to be aware of the fact that they are used as a dependency. In addition, `global` values are available globally as the name suggests and hence multiple subcharts have to have aligned value keys. As a result, the parent chart and it's dependencies are rather tightly coupled.
+
+With the `import-values` directive, there is a mechanism in Helm to use the default values of a subchart as a value in the parent chart and define the name under which the subchart value is available in the parent chart.
+This HIP describes the `export-values` directive which is similar to the [`import-values`](https://helm.sh/docs/topics/charts/#importing-child-values-via-dependencies) directive, and allows exporting values from the parent chart to the subchart and defining the name under which it should be available.
+With that the values of the parent chart and the values of the subcharts can be defined independently. Subcharts can use their local values exclusively and don't have to be aware of the usage as a subchart. Parent charts on the other hand, can provide a consistent configuration api to their users and orchestrate the exchange of values between the subcharts using `export-values`, `import-values` or a combination thereof.
+
+This change was discussed in [helm/#3242]. There were several attempts to implement this: [helm/#7477], [helm/#10059].
+
+## Specification
+
+With the proposed `export-values` directive, useres are able to specify in the **parent** chart (snippet below) which and where values should be exported to the **subchart**.
+
+```yaml
+# Chart.yaml:
+dependencies:
+  - name: client
+    version: 1.0.0
+    export-values:
+    - parent: "port"
+      child: "port"
+  - name: server
+    version: 1.0.0
+    export-values:
+    - parent: "port"
+      child: "port"
+```
+
+```yaml
+# values.yaml
+port: 8080
+```
+
+### Desired Evaluation Order
+
+1. Composite chart receives values
+   1. Explicitly set via `--set KEY=VALUE` flag(s)
+   1. Explicitly set via `--values` flag(s)
+   1. Chart defaults (`values.yaml`)
+1. Subchart receives values
+
+### Value Precedence
+
+1. Explicitly set for composit value (`port`)
+1. Explicitly set for subchart value (`client.port`)
+1. default for composit value (`values.yaml` in composite chart)
+1. default for subchart value (`values.yaml` in subchart)
```

Comment:
Should this be in fact one section about precedence? I would expect we give some quick motivation why/that value precedence is something to be considered and "spec'ed out".

Followed by where values could come from (including `import-value` and `export-values`) and a proposed precedence.

Reply:
> I would expect we give some quick motivation why/that value precedence is something to be considered and "spec'ed out".

Something like: "Because changing the order of precedence between different sources of values later would be a breaking change (same input, but potentially different output), we should agree on a user-friendly order of precedence before implementing this HIP. Suggest precedence:"

Reply:
Yes, along those lines. And, I guess it'd even make sense to visualize the potential data flows. I.e. a global visualization of parent, child (and grandchildren) + global effective values (i.e. as templates will find them in `.Values`) and where they come from (e.g. a child value directly set, but also a default mapped via export and import directives).

We need to make this as intuitive as possible and even more importantly well defined.